### DISCUSSION TITLE
How to profile application startup in a right way

### DISCUSSION THREAD

**Discussion by MikhailAsadchyIDT** [async-profiler/async-profiler#394] [Q&A] [answered]
I have a [simple command line application with micronaut](https://micronaut-projects.github.io/micronaut-picocli/latest/guide/#create). 

I found the method `io.micronaut.context.env.DefaultEnvironment#determineCloudProvider` takes majority of startup time.
It tries to do the first http call to check if app launched in GCP (`io.micronaut.context.env.DefaultEnvironment#isGoogleCompute`).

But the async-profiler says this method takes ~3% of main thread, where I'd expected it to be like 40-50%, cause it takes about 500ms while the startup time is about 1s. 

**Question:**
Could someone, please, help me to understand how I should configure async-profiler to find such most time consuming places?

**Async-profiler config, which I used:**
`-agentpath:/opt/async-profiler-1.8.4-linux-x64/build/libasyncProfiler.so=start,jfr,interval=10us,framebuf=15000000,event=wall,threads,file=/tmp/asyncprofiler489.jfr`

**Video:**

https://user-images.githubusercontent.com/50206054/110216974-a318d180-7ec2-11eb-92ec-e45f32cffb52.mp4

**Comment by apangin** [comment, marked as answer]
The config is fine, except that the minimum wall-clock profiling interval is 100us, and the "normal" interval for start-up profiling is ~1ms.

What's wrong with the results? (Or why do you think there is something wrong with them?)

**Reply by MikhailAsadchyIDT** [reply]
Thank you.
Where can I find recommended run configurations for different use cases?

**Reply by MikhailAsadchyIDT** [reply]
**What I can see**
I did a measurement of the method duration by setting two conditional break points
![Selection_099](https://user-images.githubusercontent.com/50206054/110250448-61ecf400-7f8c-11eb-8d94-5ba8ea111a75.png)

Measurement says the method duration is about 500 ms. 
There is a property to disable this method logic. If I do this, bootstrap time reduces almost twice (approximately this 500 ms). That's why I think this method duration is 500 ms.

**How I understand the profiler results**
Actually flame graph says it's about 10 samples. 
![Selection_100](https://user-images.githubusercontent.com/50206054/110250522-a7112600-7f8c-11eb-9808-3dad6d3bf341.png)

As I understand 1 sample with 1 ms interval means the total duration of the frame is 1 ms for each sample.
I know the method invocation count is 1.
That means to me flame graph says that approximate method duration is 10 ms. 
I'd expect this frame have seen in 500 samples instead of 10 as report says.

That's why I think there is something wrong with my understanding of the report. Probably I miss something.

**Reply by MikhailAsadchyIDT** [reply]
I'm trying to find the most time consuming methods. 
I found this method manually. 
For now I'm trying to understand how I can find this method with profiler.

**Reply by apangin** [reply]
Setting breakpoints in the code affects how the code is executed. In particular, JIT compilation of methods with breakpoints is completely disabled. This means, it's not quite correct to do performance measurements with breakpoints.

So far I don't see a reliable indication to think the profiler results are wrong. If you have a minimal complete verifiable example that shows a suspicious behavior, please include it.

**Reply by apangin** [reply]
I tried to repeat your experiment, and here are the results I've got:
```
$ time java -agentlib:asyncProfiler=start,event=wall,interval=1ms,threads,file=slow.html -jar /dev/shm/my-cli-app-0.1-all.jar -v
Profiling started
00:41:28.282 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [cli]
Hi!

real    0m1.229s
user    0m1.542s
sys     0m0.997s

$ time java -Dmicronaut.cloud.platform=BARE_METAL -agentlib:asyncProfiler=start,event=wall,interval=1ms,threads,file=fast.html -jar /dev/shm/my-cli-app-0.1-all.jar -v
Profiling started
00:44:47.739 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [cli]
Hi!

real    0m1.117s
user    0m1.584s
sys     0m0.739s
```

Without cloud detection, the application indeed started 112ms faster.

I then opened `slow.html` Flame Graph, zoomed into the main thread, and searched for `determineCloudProvider`. It showed that the highlighted frame took 8.47% of the main thread, which is very close to the actual measured difference (112 ms). Seems like there is no contradiction.

![slow](https://user-images.githubusercontent.com/1749416/110256154-632c1a00-7fa8-11eb-9a98-29bafd0e96a2.png)

**Reply by MikhailAsadchyIDT** [reply]
Thank you for your answers! 
Sorry, looks like the issue on IDEA side.

**Reply by MikhailAsadchyIDT** [reply]
Prof of the method slowness. 

It takes about a second on my machine

https://user-images.githubusercontent.com/50206054/111079714-3f1e8a80-850c-11eb-8928-b6f56fc1b3f5.mp4

**Reply by MikhailAsadchyIDT** [reply]
I can see this method highlighted in html flame graph (if I do from command line) and as well as in jfr with mission control, but not with IDEA report. 
It differs from what show html flame graph and mission control from jfr.

**Reply by MikhailAsadchyIDT** [reply]
From console:

https://user-images.githubusercontent.com/50206054/111079903-24004a80-850d-11eb-96bc-6e4954bf67d8.mp4

Html shows very believable picture.

So, looks like the issue is not on profiler side.

**Reply by MikhailAsadchyIDT** [reply]
When I run profiler from IDEA, it shows me flame graph, but looks like something wrong with it. It says the method only shown in 13 samples. 

But jmc shows another picture, which looks the same as html report.

https://user-images.githubusercontent.com/50206054/111080068-c0c2e800-850d-11eb-920a-74193f5d7ea0.mp4

**Reply by MikhailAsadchyIDT** [reply]
I've double check with IDEA EAP - graph is the same.

**Reply by MikhailAsadchyIDT** [reply]
Looks like IDEA integration bug with profiler.

**Reply by apangin** [reply]
Thanks for the detailed report. I think I understand the problem.
IDEA seems to display only events with threadState == STATE_RUNNABLE. However, when a thread is waiting for a network response, it is in STATE_SLEEPING.
Let me invite engineers from IDEA Profiler team for a comment.

**Reply by apangin** [reply]
Got confirmation that the issue will be fixed on IDEA side.

**Reply by MikhailAsadchyIDT** [reply]
Thank you

**Reply by gorrus** [reply]
filed https://youtrack.jetbrains.com/issue/IDEA-264335
